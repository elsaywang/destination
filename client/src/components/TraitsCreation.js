import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Add from '@react/react-spectrum/Icon/Add';
import Button from '@react/react-spectrum/Button';
import Link from '@react/react-spectrum/Link';
import styles from './TraitsCreation.css';

class TraitsCreation extends Component {
    singleRowTraitsCreation = () => {
        const { traitsCreationLabelText } = this.props;
        return (
            <Link href="#">
                <div className={styles.singleCreation}>
                    <Add size="S" />
                    <div className={styles.singleCreationlinkText}>{traitsCreationLabelText}</div>
                </div>
            </Link>
        );
    };

    multiRowsTraitsCreation = () => {
        const { handleTraitsCreation, selectedSignals, traitsCreationLabelText } = this.props;

        const { selectionMessage, warning } = selectedSignals;

        return (
            <div className={styles.multiCreation}>
                <span className={styles.message}>{selectionMessage}</span>
                <Button
                    label={traitsCreationLabelText}
                    icon={<Add />}
                    onClick={handleTraitsCreation}
                    variant="action"
                    disabled={warning}
                />
            </div>
        );
    };

    render() {
        const { multiCreation, handleTraitsCreation } = this.props;
        return multiCreation ? this.multiRowsTraitsCreation() : this.singleRowTraitsCreation();
    }
}

TraitsCreation.propTypes = {
    multiCreation: PropTypes.bool,
    selectedSignals: PropTypes.object,
    traitsCreationLabelText: PropTypes.string,
    handleTraitsCreation: PropTypes.func,
};

export default TraitsCreation;
