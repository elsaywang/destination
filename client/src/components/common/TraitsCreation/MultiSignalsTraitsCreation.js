import React from 'react';
import PropTypes from 'prop-types';
import styles from './TraitsCreation.css';
import Button from '@react/react-spectrum/Button';
import Add from '@react/react-spectrum/Icon/Add';

const MultiSignalsTraitsCreation = ({
    selectedSignals,
    storeSessionAndNavigateToTraits,
    isMaxSignalSelectionsReached,
}) => {
    const {
        selectionMessage,
        hasSignalSelectionsTypeWarning,
        hasOnboardedSignalSelectionsWarning,
    } = selectedSignals;

    const isTraitCreationDisabled =
        hasSignalSelectionsTypeWarning ||
        isMaxSignalSelectionsReached ||
        hasOnboardedSignalSelectionsWarning;

    return (
        <div className={styles.multiCreation}>
            <span className={styles.message}>{selectionMessage}</span>
            <Button
                onClick={storeSessionAndNavigateToTraits}
                label="Create Trait From Multiple Signals"
                icon={<Add />}
                variant="action"
                disabled={isTraitCreationDisabled}
                data-test="multi-signals-trait-creation"
            />
        </div>
    );
};

MultiSignalsTraitsCreation.propTypes = {
    selectedSignals: PropTypes.shape({
        selectionMessage: PropTypes.string,
        hasSignalSelectionsTypeWarning: PropTypes.bool,
        hasOnboardedSignalSelectionsWarning: PropTypes.bool,
        selectedRowIndexes: PropTypes.array,
    }),
    storeSessionAndNavigateToTraits: PropTypes.func.isRequired,
    isMaxSignalSelectionsReached: PropTypes.bool,
};

export default MultiSignalsTraitsCreation;
