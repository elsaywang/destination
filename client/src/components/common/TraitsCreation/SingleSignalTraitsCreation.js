import React from 'react';
import PropTypes from 'prop-types';
import styles from './TraitsCreation.css';
import Add from '@react/react-spectrum/Icon/Add';

const SingleSignalTraitsCreation = ({
    createTraitUrl,
    traitsCreationLabelText,
    canCreateTrait,
}) => {
    if (canCreateTrait) {
        return (
            <a className="spectrum-Link" href={createTraitUrl}>
                <div className={styles.singleCreation}>
                    <Add size="S" />
                    <div className={styles.singleCreationlinkText}>{traitsCreationLabelText}</div>
                </div>
            </a>
        );
    }

    return '0 Traits';
};

SingleSignalTraitsCreation.propTypes = {
    createTraitUrl: PropTypes.string,
    traitsCreationLabelText: PropTypes.string,
    canCreateTrait: PropTypes.bool,
};

export default SingleSignalTraitsCreation;
