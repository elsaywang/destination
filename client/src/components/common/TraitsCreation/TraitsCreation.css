:local(.singleCreationlinkText) {
    display: inline-flex;
    font-size: 0.875rem;
    padding-left: 0.15em;
    vertical-align: text-bottom;
}

:local(.singleCreation) {
    cursor: pointer;
    margin-left: -1.05em;
    white-space: nowrap;
}

:local(.multiCreation) {
    display: flex;
    justify-content: flex-end;
}

:local(.message) {
    display: flex;
    padding: 0.5em;
    font-weight: 800;
    white-space: nowrap;
}
