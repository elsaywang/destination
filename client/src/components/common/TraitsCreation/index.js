import TraitsCreation from './TraitsCreation';

export default TraitsCreation;
