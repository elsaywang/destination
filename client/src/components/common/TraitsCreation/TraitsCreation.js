import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MultiSignalsTraitsCreation from './MultiSignalsTraitsCreation';
import SingleSignalTraitsCreation from './SingleSignalTraitsCreation';
import { createOnboardedTraitUrl, createRuleBasedTraitUrl } from '../../../utils/urls';
import { stringifySignals } from '../../../utils/stringifySignals';

class TraitsCreation extends Component {
    getCreateTraitURL() {
        return this.shouldCreateOnboardedTrait()
            ? createOnboardedTraitUrl()
            : createRuleBasedTraitUrl();
    }

    getLinkText() {
        return this.props.signalType === 'ONBOARDED'
            ? 'Create Onboarded Trait'
            : 'Create Rule-Based Trait';
    }

    getButtonText() {
        return this.props.selectedResults.length > 1
            ? 'Create Trait From Selected Signals'
            : 'Create Trait From Selected Signal';
    }

    shouldCreateOnboardedTrait() {
        // For single signal creation, check the `signalType` of the signal result.
        // For multi signal creation, check if all selected signals are onboarded and
        // come from the same data source.
        const { signalType, doAllSelectedResultsShareSameDataSource } = this.props;

        return signalType === 'ONBOARDED' || doAllSelectedResultsShareSameDataSource;
    }

    storeSessionAndNavigateToTraits = e => {
        e.preventDefault();

        const { keyValuePairs, multiCreation, selectedResults, selectedDataSourceIds } = this.props;

        const signals = multiCreation ? selectedResults : [{ keyValuePairs }];
        const signalsParams = {
            signals: stringifySignals(signals),
            ...(this.shouldCreateOnboardedTrait() && {
                source: {
                    dataSourceIds: selectedDataSourceIds,
                },
            }),
        };

        sessionStorage.setItem('signalsParams', JSON.stringify(signalsParams));

        window.location.assign(this.getCreateTraitURL());
    };

    render() {
        const {
            multiCreation,
            selectedSignals,
            canCreateTraits,
            isMaxSignalSelectionsReached,
        } = this.props;

        return multiCreation ? (
            <MultiSignalsTraitsCreation
                selectedSignals={selectedSignals}
                storeSessionAndNavigateToTraits={this.storeSessionAndNavigateToTraits}
                isMaxSignalSelectionsReached={isMaxSignalSelectionsReached}
                multiTraitCreationButtonText={this.getButtonText()}
            />
        ) : (
            <SingleSignalTraitsCreation
                traitsCreationLabelText={this.getLinkText()}
                canCreateTraits={canCreateTraits}
                storeSessionAndNavigateToTraits={this.storeSessionAndNavigateToTraits}
            />
        );
    }
}

TraitsCreation.propTypes = {
    signalType: PropTypes.string,
    keyValuePairs: PropTypes.array,
    multiCreation: PropTypes.bool,
    selectedSignals: PropTypes.shape({
        selectionMessage: PropTypes.string,
        hasSignalSelectionsTypeWarning: PropTypes.bool,
        selectedRowIndexes: PropTypes.array,
        hasOnboardedSignalSelectionsWarning: PropTypes.bool,
    }),
    selectedDataSourceIds: PropTypes.array,
    canCreateTraits: PropTypes.bool,
    isMaxSignalSelectionsReached: PropTypes.bool,
};

export default TraitsCreation;
