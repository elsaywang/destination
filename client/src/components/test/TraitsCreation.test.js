import React from 'react';
import { shallow } from 'enzyme';
import TraitsCreation from '../TraitsCreation';
import Button from '@react/react-spectrum/Button';
import Add from '@react/react-spectrum/Icon/Add';
import Link from '@react/react-spectrum/Link';

describe('<TraitsCreation/> component', () => {
    describe('rendering when it is used in Single-Signal Traits Creation', () => {
        const mockFn = jest.fn();
        const props = {
            multiCreation: false,
            selectedSignals: { selectionMessage: '', warning: false },
            traitsCreationLabelText: 'Create Onboarded Trait',
            handleTraitsCreation: mockFn,
        };

        const wrapper = shallow(<TraitsCreation {...props} />);
        const creationWrapper = wrapper.instance();

        it('matches the snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('renders with prop `multiCreation` is false', () => {
            expect(creationWrapper.props.multiCreation).toEqual(false);
        });

        it('passes `traitsCreationLabelText` props with correct text', () => {
            expect(creationWrapper.props.traitsCreationLabelText).toEqual(
                props.traitsCreationLabelText,
            );
        });

        it('includes <Add/> icon with size=`S`', () => {
            expect(wrapper.find(Add).exists()).toBe(true);
            expect(wrapper.find(Add).props().size).toEqual('S');
        });

        it('renders the correct text passed from `traitsCreationLabelText` prop in a <div/>', () => {
            expect(
                wrapper.containsMatchingElement(<div>{props.traitsCreationLabelText}</div>),
            ).toBeTruthy();
        });
    });

    describe('rendering when it is used in Multi-Signals Traits Creation', () => {
        const mockFn = jest.fn();
        const props = {
            multiCreation: true,
            selectedSignals: { selectionMessage: '', warning: false },
            traitsCreationLabelText: 'Create Trait From Multi Signals',
            handleTraitsCreation: mockFn,
        };

        const wrapper = shallow(<TraitsCreation {...props} />);
        const creationWrapper = wrapper.instance();

        it('matches the snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('renders with prop `multiCreation` is true', () => {
            expect(creationWrapper.props.multiCreation).toEqual(true);
        });

        it('passes `traitsCreationLabelText` props with correct text', () => {
            expect(creationWrapper.props.traitsCreationLabelText).toEqual(
                props.traitsCreationLabelText,
            );
        });

        it('renders the correct message passed from `selectedSignals.selectionMessage` prop in a <span/>', () => {
            expect(
                wrapper.containsMatchingElement(
                    <span>{props.selectedSignals.selectionMessage}</span>,
                ),
            ).toBeTruthy();
        });

        it('renders <Button /> component has correct labeling passed from `traitsCreationLabelText` prop', () => {
            expect(wrapper.find(Button).exists()).toBe(true);
            expect(wrapper.find(Button).props().label).toEqual(props.traitsCreationLabelText);
        });

        it('renders <Button /> component includes <Add/> as an icon prop', () => {
            expect(wrapper.find(Button).props().icon).toEqual(<Add />);
        });

        it('renders non-disabled <Button /> with "action" variant when no warning message passes from props', () => {
            expect(wrapper.find(Button).props().variant).toEqual('action');
            expect('disabled' in wrapper.props()).toBeFalsy();
        });

        it('renders disabled <Button /> with "action" variant when warning message passes from props', () => {
            const propsWithWarning = {
                selectedSignals: {
                    selectionMessage: '1 Real-time signal , 1 Onboarded signal selected ',
                    records: [
                        { rowIndex: 0, signalType: 'Adobe Analytics' },
                        { rowIndex: 1, signalType: 'Onboarded Records' },
                    ],
                    warning: true,
                },
            };
            wrapper.setProps({ ...propsWithWarning });
            expect(wrapper.find(Button).props().variant).toEqual('action');
            expect(wrapper.find(Button).props().disabled).toBe(true);
        });
    });
});
