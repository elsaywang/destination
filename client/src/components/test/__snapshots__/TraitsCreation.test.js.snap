// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TraitsCreation/> component rendering when it is used in Multi-Signals Traits Creation matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TraitsCreation
    handleTraitsCreation={[MockFunction]}
    multiCreation={true}
    selectedSignals={
      Object {
        "selectionMessage": "",
        "warning": false,
      }
    }
    traitsCreationLabelText="Create Trait From Multi Signals"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span
          className={undefined}
        >
          
        </span>,
        <Button
          block={false}
          disabled={false}
          element="button"
          icon={<Add />}
          invalid={false}
          label="Create Trait From Multi Signals"
          logic={false}
          onClick={[MockFunction]}
          quiet={false}
          selected={false}
          variant="action"
        />,
      ],
      "className": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": undefined,
        },
        "ref": null,
        "rendered": "",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "block": false,
          "disabled": false,
          "element": "button",
          "icon": <Add />,
          "invalid": false,
          "label": "Create Trait From Multi Signals",
          "logic": false,
          "onClick": [MockFunction],
          "quiet": false,
          "selected": false,
          "variant": "action",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span
            className={undefined}
          >
            
          </span>,
          <Button
            block={false}
            disabled={false}
            element="button"
            icon={<Add />}
            invalid={false}
            label="Create Trait From Multi Signals"
            logic={false}
            onClick={[MockFunction]}
            quiet={false}
            selected={false}
            variant="action"
          />,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": undefined,
          },
          "ref": null,
          "rendered": "",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "block": false,
            "disabled": false,
            "element": "button",
            "icon": <Add />,
            "invalid": false,
            "label": "Create Trait From Multi Signals",
            "logic": false,
            "onClick": [MockFunction],
            "quiet": false,
            "selected": false,
            "variant": "action",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<TraitsCreation/> component rendering when it is used in Single-Signal Traits Creation matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TraitsCreation
    handleTraitsCreation={[MockFunction]}
    multiCreation={false}
    selectedSignals={
      Object {
        "selectionMessage": "",
        "warning": false,
      }
    }
    traitsCreationLabelText="Create Onboarded Trait"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className={undefined}
      >
        <Add
          size="S"
        />
        <div
          className={undefined}
        >
          Create Onboarded Trait
        </div>
      </div>,
      "href": "#",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Add
            size="S"
          />,
          <div
            className={undefined}
          >
            Create Onboarded Trait
          </div>,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "size": "S",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Create Onboarded Trait",
            "className": undefined,
          },
          "ref": null,
          "rendered": "Create Onboarded Trait",
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className={undefined}
        >
          <Add
            size="S"
          />
          <div
            className={undefined}
          >
            Create Onboarded Trait
          </div>
        </div>,
        "href": "#",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Add
              size="S"
            />,
            <div
              className={undefined}
            >
              Create Onboarded Trait
            </div>,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "size": "S",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Create Onboarded Trait",
              "className": undefined,
            },
            "ref": null,
            "rendered": "Create Onboarded Trait",
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
